contracts/v0.7/bridge/optimism/OVM_L1ERC20Gateway.sol

```diff
@@ -0,0 +1,157 @@
+ *
+ * Compiler used: solc
+ * Runtime target: EVM
+ */
+contract OVM_L1ERC20Gateway is OpUnsafe, Initializable, Abs_L1TokenGateway {
+  // L1 token we are bridging to L2
+  IERC20 public s_l1ERC20;
+
+  // This contract lives behind a proxy, so the constructor parameters will go unused.
+  constructor()
+    Abs_L1TokenGateway(
+      address(0), // _l2DepositedToken
+      address(0) // _l1messenger
+    )
+    public
+  {}
+
+  /**
+   * @param l2ERC20Gateway L2 Gateway address on the chain being deposited into
+   * @param l1Messenger Cross-domain messenger used by this contract.
+   * @param l1ERC20 L1 ERC20 address this contract stores deposits for
+   */
+  function init(
+    address l2ERC20Gateway,
+    address l1Messenger,
+    IERC20 l1ERC20
+  )
+    initializer()
+    public
+  {
+    // TODO: require != address(0)?
+    s_l1ERC20 = l1ERC20;
+    // Init parent contracts
+    l2DepositedToken = l2ERC20Gateway;
+    messenger = l1Messenger;
+  }
+
+  /// @dev Modifier requiring the contract to be initialized
+  modifier onlyInitialized() {
+    require(address(l2DepositedToken) != address(0), "Contract not initialized");
+    _;
+  }
+
+  /// @dev Returns L2 ERC20 Gateway address (AKA l2DepositedToken).
+  function l2ERC20Gateway()
+    public
+    view
+    returns (address)
+  {
+    // Default Optimism ERC20 bridge implemenation combines the L2 gateway and token
+    // into a single OVM_L2DepositedERC20 contract. From the perspective of L1 gateway,
+    // this should be just an implementation detail, so here we expose an address in a
+    // different more general name "l2ERC20Gateway".
+    return l2DepositedToken;
+  }
+
+
+  /**************
+   * Depositing *
+   **************/
+
+  /**
+   * @dev deposit an amount of ERC20 to a recipients's balance on L2
+   * WARNING: This is a potentially unsafe operation that could end up with lost tokens,
+   * if tokens are sent to a contract. Be careful!
+   *
+   * @param _to L2 address to credit the withdrawal to
+   * @param _amount Amount of the ERC20 to deposit
+   */
+  function depositToUnsafe(
+    address _to,
+    uint _amount
+  )
+    public
+    unsafe()
+  {
+    _initiateDeposit(msg.sender, _to, _amount);
+  }
+
+
+  /**************
+   * Accounting *
+   **************/
+
+  /**
+   * @dev When a deposit is initiated on L1, the L1 Gateway
+   * transfers the funds to itself for future withdrawals
+   *
+   * @param _from L1 address ERC20 is being deposited from
+   * @param _to L2 address that the ERC20 is being deposited to
+   * @param _amount Amount of ERC20 to send
+   */
+  function _handleInitiateDeposit(
+    address _from,
+    address _to,
+    uint256 _amount
+  )
+    internal
+    override
+    onlyInitialized()
+  {
+    // Unless explicitly unsafe op, stop deposits to contracts (avoid accidentally lost tokens)
+    require(_isUnsafe() || !Address.isContract(_to), "Unsafe deposit to contract");
```

Comment:
I'm a bit confused by how this line functions. Since this contract exists on L1, but the `_to` address is pointing at an L2 address. How can `Address.isContract(_to)` check if it's a contract on L2 from L1?
(this might have to do with my lack of understanding of how Optimism fully works...)

Reply:
Wait...I think I figured it out. Correct me if I am wrong:
* Since OVM is a roll up solution where it is completely dependent on L1, the addresses line up directly with what is seen on L1
* So if the address on L1 is a smart contract, funds sent to that address on L2 would only be accessible if the smart contract also had integrated with OVM

Reply:
The questions are great! :)

This specific line is a little bit confusing, mostly because we could not extend `deposit/depositTo` & `withdraw/withdrawTo` (they are not virtual). So I added an `unsafe` marker to the third function `depositToUnsafe/withdrawToUnsafe` to be able to set up this check.

**We want to block the bridge transfers to (local) contracts, unless explicit.**

This is because, when sending to a contract it's easy to get tokens stuck, as you need to have a contract be available at the same address on L2. The "only way" to have the same contract address on L1 & L2 is to use CREATE with the same nonce. The CREATE2 opcode will not work, as the OVM compiled bytecode will be different than EVM (for contracts of any relevant complexity), so they will get different addresses on L1 & L2 networks.

This actually happened to a Synthetix user with an Argent SC wallet, that used the bridge from his smart wallet and locked his funds. The wallet sent to itself on L2, and there is no way (except luck) to recreate the same address on L2.

Reply:
That's a great explanation! And yes blocking token transfers to contracts makes sense. 

If I understand your explanation correctly, then on L1 there could be a smart contract deployed at a certain address, but not on L2 (and vice versa). 

So I think I still have my original question: how can a contract deployed on L1 as the gateway detect that an address on L2 is a smart contract or not? it seems like `!Address.isContract(_to)` would only check on L1...

Reply:
Yes, we only check for local contracts, L1 gateway checks for L1 contracts, and similarly on L2.

If an L1 gateway receives a transfer request to L2 address, it checks if that address is a contract on L1. If it is, the gateway makes sure the user is explicit in its intention. If not, the gateway rejects to avoid accidental mistakes, as the contract address space, at least for the CREATE2 opcode, is vastly different for the two networks.

Reply:
Cool, that makes sense! thanks :)